### ISSUE TITLE
Defect: cafrun fails with >=48 processes

### ISSUE THREAD

**Issue by lstagner** [sourceryinstitute/OpenCoarrays#496]
[issue stat img]: https://img.shields.io/issuestats/i/github/sourceryinstitute/OpenCoarrays.svg?style=flat-square

| Avg response time              |
|:------------------------------:|
| ![Issue Stats][issue stat img] |

# Defect/Bug Report #

- OpenCoarrays Version: `1.9.3`
- Fortran Compiler: `gfortran 7.1`
- C compiler used for building lib: `gcc 7.1`
- Installation method:`install.sh`
- Output of `uname -a`: `Linux irisa.cluster 2.6.32-696.3.2.el6.x86_64 #1 SMP Tue Jun 20 01:26:55 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`
- MPI library being used: MPICH 3.2
```
[stagnerl@irisa deps]$ cafrun -w
HYDRA build details:
    Version:                                 3.2
    Release Date:                            Wed Nov 11 22:06:48 CST 2015
    CC:                              gcc    
    CXX:                             g++    
    F77:                             gfortran   
    F90:                             gfortran   
    Configure options:                       '--disable-option-checking' '--prefix=/fusion/projects/codes/gcc71-toolchain' 'CC=gcc' 'CXX=g++' 'F77=gfortran' 'FC=gfortran' '--enable-fast=all,O3' '--enable-debuginfo' '--with-pm=hydra' '--cache-file=/dev/null' '--srcdir=../../../../mpich-3.2/src/pm/hydra' 'CFLAGS= -DNDEBUG -DNVALGRIND -O3' 'LDFLAGS=' 'LIBS=-lpthread ' 'CPPFLAGS= -I/cluster-scratch/halpernf/source/mpich_build/src/mpl/include -I/cluster-scratch/halpernf/source/mpich-3.2/src/mpl/include -I/cluster-scratch/halpernf/source/mpich-3.2/src/openpa/src -I/cluster-scratch/halpernf/source/mpich_build/src/openpa/src -D_REENTRANT -I/cluster-scratch/halpernf/source/mpich_build/src/mpi/romio/include'
    Process Manager:                         pmi
    Launchers available:                     ssh rsh fork slurm ll lsf sge manual persist
    Topology libraries available:            hwloc
    Resource management kernels available:   user slurm ll lsf sge pbs cobalt
    Checkpointing libraries available:       
    Demux engines available:                 poll select

```
- Machine architecture and number of physical cores: x86_64 &  16

## Observed Behavior ##
cafrun fails with processes >48. It runs fine with 47 processes but not with 48
```
Fatal error in PMPIX_Comm_agree: Unknown error class, error stack:
PMPIX_Comm_agree(183).........: MPIX_Comm_agree(MPI_COMM_WORLD) failed
PMPIX_Comm_agree(169).........: 
MPIR_Comm_agree(98)...........: MPIX_Comm_agree failed
MPIR_Allreduce_group(377).....: 
MPIR_Allreduce_group_intra(53): Invalid rank
Fatal error in PMPIX_Comm_agree: Unknown error class, error stack:
PMPIX_Comm_agree(183).........: MPIX_Comm_agree(MPI_COMM_WORLD) failed
PMPIX_Comm_agree(169).........: 
MPIR_Comm_agree(98)...........: MPIX_Comm_agree failed
MPIR_Allreduce_group(377).....: 
MPIR_Allreduce_group_intra(53): Invalid rank
Fatal error in PMPIX_Comm_agree: Unknown error class, error stack:
PMPIX_Comm_agree(183).........: MPIX_Comm_agree(MPI_COMM_WORLD) failed
PMPIX_Comm_agree(169).........: 
MPIR_Comm_agree(98)...........: MPIX_Comm_agree failed
MPIR_Allreduce_group(377).....: 
MPIR_Allreduce_group_intra(53): Invalid rank
Fatal error in PMPIX_Comm_agree: Unknown error class, error stack:
PMPIX_Comm_agree(183).........: MPIX_Comm_agree(MPI_COMM_WORLD) failed
PMPIX_Comm_agree(169).........: 
MPIR_Comm_agree(98)...........: MPIX_Comm_agree failed
MPIR_Allreduce_group(377).....: 
MPIR_Allreduce_group_intra(53): Invalid rank
Fatal error in PMPIX_Comm_agree: Unknown error class, error stack:
PMPIX_Comm_agree(183).........: MPIX_Comm_agree(MPI_COMM_WORLD) failed
PMPIX_Comm_agree(169).........: 
MPIR_Comm_agree(98)...........: MPIX_Comm_agree failed
MPIR_Allreduce_group(377).....: 
MPIR_Allreduce_group_intra(53): Invalid rank
Error in /home/stagnerl/FIDASIM/deps/opencoarrays/bin/cafrun in function . on line 265. Please report this error at http://bit.ly/OpenCoarrays-new-issue
```
## Steps to Reproduce ##
?

**Comment by zbeekman** [comment]
@lstagner Thanks for the bug report.

Can you please show me:

```
caf --show
cafrun --show
```

Also, does this happen with every CAF code, or just a particular program? Any chance of being able to see the source code that triggered this, or a MWE/MCVE?

This is certainly odd. Can you run MPI programs oversubscribed at 3:1 without problems?

I know there is a bug in MPICH 3.2 (I think it's also still in 3.2.1 and maybe even the 3.3 preview release) that has detrimental effects when the OC library is compiled with failed image support. The patch to fix it is here: https://github.com/pmodels/mpich/pull/2730 If this bug is related to this issue, then patching MPICH may fix it.

Also, you may want to try disabling failed image support, since that depends on experimental user level fault mitigation features proposed for the next MPI standard. Building through CMake directly:

1. Make sure `mpiexec` or `mpirun` is on your PATH for the MPI implementation you wish to use
2. `export FC=/path/to/your/gfortran && export CC=/path/to/your/C/compiler` or whatever the equivalent is for your shell. We recommend GCC matching the version of GFortran for the C compiler, but clang works most of the time too
3. `mkdir build && cd build`
4. `cmake /path/to/OC-source-dir -DCMAKE_INSTALL_PREFIX=/where/to/install/OC/libs -DCAF_ENABLE_FAILED_IMAGES=FALSE`
5. `make -j install`
6. `make check # optional but recommended`
7. To uninstall `make uninstall`

**Comment by zbeekman** [comment]
UPDATE: After actually reading the error messages, rather than just skimming them, it appears that this is definitely caused by either buggy support for ULFM experimental MPI features (the MPIX_... calls) or a bug in OpenCoarrays failed image implementation.

If you don't need failed images, the best thing to do is probably to just build directly with CMake, as outlined above, turning off failed images support.

If you'd rather use the build from the install script, you can also enter the `./prerequisites/builds/opencoarrays/<version>/` and edit the CMake cache to turn off failed images support via `ccmake ../../../..` and then going to the `CAF_ENABLE_FAILED_IMAGES` entry, toggling it off via <kbd>Enter</kbd> then pressing <kbd>c</kbd> to reconfigure (dismiss any developer warnings with <kbd>e</kbd>) and <kbd>g</kbd> to regenerate and quit `ccmake`. (If you have `cmake-gui` you can accomplish the same thing via `cmake-gui ../../../..` and then using the GUI.) You could also run `cmake ../../../.. -DCAF_ENABLE_FAILED_IMAGES=FALSE`. Then `make clean && make -j install && make check`.

If you want/need failed image support, please try building MPICH from the master branch, or patching 3.2 with the patch at https://github.com/pmodels/mpich/pull/2730.patch

**Comment by lstagner** [comment]
```
[stagnerl@iris26 ~]$ caf --show
/fusion/projects/codes/gcc71-toolchain/bin/gfortran -I/home/stagnerl/FIDASIM/deps/opencoarrays/include/OpenCoarrays-1.9.3_GNU-7.1.0 -fcoarray=lib -Wl,-rpath -Wl,/fusion/projects/codes/gcc71-toolchain/lib -Wl,--enable-new-dtags ${@} /home/stagnerl/FIDASIM/deps/opencoarrays/lib64/libcaf_mpi.a /fusion/projects/codes/gcc71-toolchain/lib/libmpifort.so /fusion/projects/codes/gcc71-toolchain/lib/libmpi.so
[stagnerl@iris26 ~]$ cafrun --show
/fusion/projects/codes/gcc71-toolchain/bin/mpirun -np <number_of_images> --disable-auto-cleanup /path/to/coarray_Fortran_program [arg4 [arg5 [...]]]
```

Playing around with it today the number of processes I can use has changed around. I presume it is related to the state of the cluster. 

I have also tried allocating a single node with 16 processors and I can use up to 65 processes and then it causes a new error.
```
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Segmentation fault (signal 11)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
Error in /home/stagnerl/FIDASIM/deps/opencoarrays/bin/cafrun in function . on line 265. Please report this error at http://bit.ly/OpenCoarrays-new-issue
``` 
This error also happens with a different program.

I'll try disabling failed image support.

You can see view the branch where I am doing the work here
https://github.com/D3DEnergetic/FIDASIM/tree/ls/coarrays2

**Comment by lstagner** [comment]
What is the consequences of not using failed image support

**Comment by zbeekman** [comment]
Failed image support is a feature of Fortran 2018 that is designed for exascale simulations. The idea is that once you have enough processing elements, you are essentially guaranteed to encounter failures. It allows the programmer to attempt to gracefully recover from such a failure, by writing logic to sideline broken nodes/PEs, and possibly swap in idle nodes/PEs included in the batch job reservation. To be truly useful, support for Fortran 2018 teams is required. Preliminary teams support should be in the GFortran 8 release with the latest and greatest OC library at that time.

If you are not explicitly writing code with these MPI experimental ULFM calls, or to check for and handle coarray failed images via Fortran 2018 `stat_failed_image` etc. then the behavior of your program should be as anticipated. FYI, if you compile with OpenMPI the build system detects that the required MPIX features are missing and automatically disables failed images support.

**Comment by zbeekman** [comment]
To get an idea, look at the directory containing the failed images tests. Right there are no tests showing how to recover, just detection of stopped and failed images: https://github.com/sourceryinstitute/OpenCoarrays/tree/1.9.3/src/tests/unit/fail_images

**Comment by lstagner** [comment]
After disabling failed image support I get a new error.
```
[proxy:0:1@iris16.cluster] HYDU_sock_write (../../../../mpich-3.2/src/pm/hydra/utils/sock/sock.c:286): write error (Bad file descriptor)
[proxy:0:1@iris16.cluster] handle_pmi_response (../../../../mpich-3.2/src/pm/hydra/pm/pmiserv/pmip_cb.c:401): unable to forward PMI response to MPI process
[proxy:0:1@iris16.cluster] HYD_pmcd_pmip_control_cmd_cb (../../../../mpich-3.2/src/pm/hydra/pm/pmiserv/pmip_cb.c:909): unable to handle PMI response
[proxy:0:1@iris16.cluster] HYDT_dmxu_poll_wait_for_event (../../../../mpich-3.2/src/pm/hydra/tools/demux/demux_poll.c:76): callback returned error status
[proxy:0:1@iris16.cluster] main (../../../../mpich-3.2/src/pm/hydra/pm/pmiserv/pmip.c:206): demux engine error waiting for event

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 18550 RUNNING AT iris10
=   EXIT CODE: 11
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
[proxy:0:2@iris17.cluster] HYD_pmcd_pmip_control_cmd_cb (../../../../mpich-3.2/src/pm/hydra/pm/pmiserv/pmip_cb.c:885): assert (!closed) failed
[proxy:0:2@iris17.cluster] HYDT_dmxu_poll_wait_for_event (../../../../mpich-3.2/src/pm/hydra/tools/demux/demux_poll.c:76): callback returned error status
[proxy:0:2@iris17.cluster] main (../../../../mpich-3.2/src/pm/hydra/pm/pmiserv/pmip.c:206): demux engine error waiting for event
[proxy:0:3@iris18.cluster] HYD_pmcd_pmip_control_cmd_cb (../../../../mpich-3.2/src/pm/hydra/pm/pmiserv/pmip_cb.c:885): assert (!closed) failed
[proxy:0:3@iris18.cluster] HYDT_dmxu_poll_wait_for_event (../../../../mpich-3.2/src/pm/hydra/tools/demux/demux_poll.c:76): callback returned error status
[proxy:0:3@iris18.cluster] main (../../../../mpich-3.2/src/pm/hydra/pm/pmiserv/pmip.c:206): demux engine error waiting for event
[proxy:0:4@iris19.cluster] HYD_pmcd_pmip_control_cmd_cb (../../../../mpich-3.2/src/pm/hydra/pm/pmiserv/pmip_cb.c:885): assert (!closed) failed
[proxy:0:4@iris19.cluster] HYDT_dmxu_poll_wait_for_event (../../../../mpich-3.2/src/pm/hydra/tools/demux/demux_poll.c:76): callback returned error status
[proxy:0:4@iris19.cluster] main (../../../../mpich-3.2/src/pm/hydra/pm/pmiserv/pmip.c:206): demux engine error waiting for event
[proxy:0:5@iris20.cluster] HYD_pmcd_pmip_control_cmd_cb (../../../../mpich-3.2/src/pm/hydra/pm/pmiserv/pmip_cb.c:885): assert (!closed) failed
[proxy:0:5@iris20.cluster] HYDT_dmxu_poll_wait_for_event (../../../../mpich-3.2/src/pm/hydra/tools/demux/demux_poll.c:76): callback returned error status
[proxy:0:5@iris20.cluster] main (../../../../mpich-3.2/src/pm/hydra/pm/pmiserv/pmip.c:206): demux engine error waiting for event
srun: error: iris16: task 1: Exited with exit code 5
srun: error: iris19: task 4: Exited with exit code 7
srun: error: iris17: task 2: Exited with exit code 7
srun: error: iris20: task 5: Exited with exit code 7
srun: error: iris18: task 3: Exited with exit code 7
[mpiexec@iris10.cluster] HYDT_bscu_wait_for_completion (../../../../mpich-3.2/src/pm/hydra/tools/bootstrap/utils/bscu_wait.c:76): one of the processes terminated badly; aborting
[mpiexec@iris10.cluster] HYDT_bsci_wait_for_completion (../../../../mpich-3.2/src/pm/hydra/tools/bootstrap/src/bsci_wait.c:23): launcher returned error waiting for completion
[mpiexec@iris10.cluster] HYD_pmci_wait_for_completion (../../../../mpich-3.2/src/pm/hydra/pm/pmiserv/pmiserv_pmci.c:218): launcher returned error waiting for completion
[mpiexec@iris10.cluster] main (../../../../mpich-3.2/src/pm/hydra/ui/mpich/mpiexec.c:344): process manager error waiting for completion
Error in /home/stagnerl/FIDASIM/deps/opencoarrays/bin/cafrun in function . on line 265. Please report this error at http://bit.ly/OpenCoarrays-new-issue
```
again it only happens when I use a large number of processes e.g. 70 but not 60

**Comment by zbeekman** [comment]
@lstagner Can I please see the new output of `cafrun` just to make sure everything was adjusted correctly? Also, are you still running on 16 physical cores, but oversubscribed with ~70 CAF images/MPI ranks? The errors look like problems in MPICH to me. Tough to know for sure without further investigation or a MWE/MCVE.

**Comment by lstagner** [comment]
```
[stagnerl@iris26 ~]$ cafrun

 cafrun - Coarray Fortran executable launcher for OpenCoarrays

 Usage: cafrun [options] ...

 Options:
   --help, -h               Show this help message
   --version, -v, -V        Report version and copyright information
   --wraps, -w,             Report info about the wrapped MPI launcher
   -np <N>,                 Number of images, N, to execute, N must be a positive integer
   --reenable-auto-cleanup  Turn off failed images support (if library support is present)
                            This option re-enables MPI auto cleanup, which is disabled by
                            by default if GFortran/OpenCoarrays/MPI all support failed
                            images through MPI ULFM. When MPI auto cleanup is disabled and
                            failed image support is present, OpenCoarrays triggers cleanup
                            explicitly when a failed/stopped image is encountered in an
                            image control statement without a `stat=` clause.
   --show, -s,              Show the command that the wrapper will execute. You can pass
                            this as the first argument and then the additional arguments
                            that you're planning to pass to perform a dry run.

 Example usage:

   cafrun -np 2 foo foo_arg1 foo_arg2
   cafrun -v
   cafrun --help
   cafrun --show
   cafrun -s -np 4 my_exe
   cafrun -np 4 --reenable-auto-cleanup ./my_exe arg1 arg2

 Notes:
   [options] must be a CAF executable file, one of the above arguments,
   or an argument to the program name returned by caf --wraps
```
The error occurs whether I am on a dedicated node or whether I submit through the queue. The only difference is that the dedicated node has a shorter message.
```
[stagnerl@iris26 ~]$ cafrun -np 70 fidasim TEST_dev/test_1b_inputs.dat 

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 144056 RUNNING AT iris26
=   EXIT CODE: 11
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Segmentation fault (signal 11)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions
Error in /home/stagnerl/FIDASIM/deps/opencoarrays/bin/cafrun in function . on line 265. Please report this error at http://bit.ly/OpenCoarrays-new-issue
```

**Comment by zbeekman** [comment]
@lstagner Sorry, I spoke too quickly. What I meant to ask for was:

1. The output of `cafrun --show` (or better yet, `cafrun --show -np 70 fidasim TEST_dev/test_1b_inputs.dat`)
2. Are there still only 16 physical cores on the node where you're executing with -np 70?

If the answer to 2. is "yes" does this happen when you don't oversubscribe? (i.e. 1 PE/core per rank)

Do you anticipate running simulations in production this oversubscribed?

Thanks for the rapid feedback!

**Comment by lstagner** [comment]
```
cafrun --show -np 70 fidasim TEST_dev/test_1b_inputs.dat 
/fusion/projects/codes/gcc71-toolchain/bin/mpirun -np 70 fidasim TEST_dev/test_1b_inputs.dat
```
> Do you anticipate running simulations in production this oversubscribed?
In practice no but the error happens when I just submit it to the queue with the following slurm jobscript
```
#!/bin/bash
#SBATCH -J fidasim
#SBATCH --ntasks=70
#SBATCH --mem-per-cpu=5G
#SBATCH -t 10:00:00
#SBATCH -o /dev/null -e /dev/null
#SBATCH -p preemptable
ulimit -s unlimited
cafrun -np 70 /home/stagnerl/FIDASIM/fidasim /home/stagnerl/TEST_dev/test_1b_inputs.dat >& /home/stagnerl/TEST_dev/test_1b.out
exit
```
Could it be possible that slurm is oversubscribing the job. Is there a way to turn it off.

**Comment by zbeekman** [comment]
Great, it looks like the rebuild of OC had the intended effect of disabling failed image support, and is now handling mpirun in the normal way. I'll see if I can determine if SLURM is oversubscribing your job with that script. Are you running on a UC Irvine machine or a nat'l lab somewhere? (I ask because I may have an account there if it is the latter.)

**Comment by lstagner** [comment]
I am running on the Iris cluster at General Atomics. 

I expicitly set the `cpus-per-task` slurm option to limit it to 1 task per cpu and it still doesn't work.

I don't understand why its saying my program is segfaulting when it doesn't with fewer processes. I even tried stopping the program after the use statements and it still segfaults.

**Comment by zbeekman** [comment]
OK, it looks like we'll have to try running it ourselves. Is the input file test_1b_inputs.dat available with the FIDASIM source code?

**Comment by lstagner** [comment]
Generating the test inputs can be a bit involved here are the all the input files you need to run FIDASIM
https://drive.google.com/open?id=12fxOUunB00PqgLMFKYpUvxPfTDnzdC35

You will need to compile FIDASIM with the option e.g. `make USE_MPI=y` it will install hdf5 and opencoarrays in the `deps` directory

**Comment by rouson** [comment]
@lstagner Are you able to post some details about the Iris cluster?  Who is the manufacturer?  In the old days (the '90s), "Iris" was a brand name for a workstation manufactured by SGI.  I'm guessing it's just a coincidence that your cluster is named Iris, but if it happens to be an SGI platform, I'll mention that we've seen a lot of problems with the advanced, MPI-3 one-sided communication features on SGI platforms.  In the case of a [paper] I co-authored, we had to switch to using the OpenSHMEM transport layer that OpenCoarrays provides. The [OpenSHMEM] layer is experimental and limited so it doesn't go into the OpenCoarrays releases, but is available via git clone.  However, in our case, the problems were with SGI's MPI implementation (MPT).  I don't recall whether we tried MPICH.

[paper]: https://bit.ly/coarray-icar-paw17
[OpenSHMEM]: https://github.com/sourceryinstitute/OpenCoarrays/tree/master/src/openshmem

**Comment by lstagner** [comment]
@rouson As far as I know the name Iris is coincidental. I actually do not know very much about the hardware other than its a 64-bit Linux cluster running CentOS 6.x and it has 30 compute nodes with 16 processors each. I don't know anything about the communication layers.

**Comment by rouson** [comment]
@lstagner Reading through this thread again, I'm trying to understand how to reconcile the following two comments:

"I have also tried allocating a single node with 16 processors and I can use up to 65 processes and then it causes a new error."

and

"I expicitly set the cpus-per-task slurm option to limit it to 1 task per cpu and it still doesn't work."

Are you confident that you're executing one image (which under the hood is 1 MPI rank) per core?  I have also seen intermittent test failures recently (mostly but not exclusively with the failed images tests), but I only see the problems when I run our tests in parallel via `ctest -j 4` inside a virtual machine that is set up to use 4 cores.  In other words, I only see the problems when I oversubscribe the cores on my platform, running more images than the available cores.  If your experience is the same, then I'm thinking we might need to mark this as won't fix and update our documentation to indicate that running more images than cores is unsupported.

**Comment by lstagner** [comment]
You can go ahead and close this. Having been using the Coarray version of the code for a little while now its not really a big problem. As you mentioned the problem only seems to show up when I oversubscribe or/and when i'm using a certain unlimited queue on a different cluster. So long as I dont oversubscribe and don't use that queue then I don't have any problems.

**Comment by zbeekman** [comment]
@lstagner thanks for the update. We have added a [FAQ] that has a number of points that are relevant for running with OpenMPI. Issues relevant to the use of a hostfile may also apply to MPICH/MVAPICH and oversubscribing issues could be due to the configuration of the cluster or limitations in the MPI implementation.

If you have any evidence that there is a race condition in the OpenCoarrays library or some other type of defect that is causing you problems when oversubscribing, please let us know.

Thanks!

[FAQ]: https://github.com/sourceryinstitute/OpenCoarrays/blob/master/FAQ.md
